microcosm_pubsub/codecs.py

```diff
@@ -99,10 +78,16 @@ def encode(self, dct=None, **kwargs):
         """
         message = dct.copy() if dct else dict()
         message.update(kwargs)
-        # We allow unknown fields to pass through (but not included here)
-        # to accommodate unconditional parameter passing during produce()
-        # e.g. passing in `uri` for IdentityMessages
-        return dumps(self.schema.load(message, unknown=EXCLUDE))
+
+        dumped = self.schema.dump(message)
+
+        # NB: Schema dump doesn't run validation
+        errors = self.schema.validate(dumped)
```

Comment:
This bit is a bit surprising, the fact that we have to run `validate` on `dumped` rather than `message`

Reply:
They're both dicts. The only difference is that the `message` is the `object` form, and `dumped` is the "pre-serialized" form (e.g. all enums cast to str, all fields converted to the wire form).

Running validate on `message` will error out, because the fields haven't yet been converted by the dump.

`validate` essentially ensures that you didn't forget to pass in data that needed to be there (e.g. didn't supply a required field).